src/AI/ScriptState.cpp

```diff
@@ -0,0 +1,591 @@
+#include "ScriptState.hpp"
+#include <RTTI/RTTI_ScriptState.hpp>
+#include <Scene/BsSceneObject.h>
+#include <components/Character.hpp>
+#include <components/CharacterAI.hpp>
+#include <components/CharacterEventQueue.hpp>
+#include <components/GameClock.hpp>
+#include <components/GameWorld.hpp>
+#include <scripting/ScriptVMForGameWorld.hpp>
+
+namespace REGoth
+{
+  namespace AI
+  {
+    /**
+     * Valid script-states an NPC can be in.
+     */
+    const bs::String s_EnabledPlayerStates[] = {
+        "ZS_ASSESSMAGIC", "ZS_ASSESSSTOPMAGIC", "ZS_MAGICFREEZE", "ZS_WHIRLWIND",
+        "ZS_SHORTZAPPED", "ZS_ZAPPED",          "ZS_PYRO",        "ZS_MAGICSLEEP",
+    };
+
+    /**
+     * Names for the native states passed to the scripts.
+     */
+    const bs::String s_NativeStateNames[] = {
+        "INVALID_PRG", "ZS_ANSWER", "ZS_DEAD", "ZS_UNCONSCIOUS", "ZS_FADEAWAY", "ZS_FOLLOW",
+    };
+
+    ScriptState::ScriptState(::REGoth::HGameWorld world, ::REGoth::HCharacter hostCharacter,
+                             ::REGoth::HCharacterEventQueue hostEventQueue,
+                             ::REGoth::HCharacterAI hostAI)
+        : mWorld(world)
+        , mHostCharacter(hostCharacter)
+        , mHostEventQueue(hostEventQueue)
+        , mHostAI(hostAI)
+    {
+    }
+
+    ScriptState::~ScriptState()
+    {
+    }
+
+    void ScriptState::interruptActiveState()
+    {
+      mCurrentState.phase = AIState::Phase::Interrupt;
+
+      // Set to invalid so the queued state can take over in the next cycle
+      mCurrentState.isValid = false;
+
+      // Non-Player Characters can use any state. But the Player character has only
+      // a limited subset. It is unclear who would start a state that is invalid on
+      // the player however. Also, shouldn't that check be done much earlier?
+      if (!mHostCharacter->isPlayer() || canPlayerUseAIState(mNextState))
+      {
+        // FIXME: I think we need to also kill the active route the character is traveling on
+        mHostEventQueue->clear();
+      }
+    }
+
+    void ScriptState::requestEndActiveState()
+    {
+      // Let the state exit normally over the next cycles
+      mCurrentState.phase = AIState::Phase::End;
+    }
+
+    void ScriptState::startNativeAIState(NativeState state)
+    {
+      bs::UINT32 index = (bs::UINT32)state;
```

Comment:
May want to use C++ style cast here?

Reply:
IMHO C-Style casts are nicer for simple casts between integer types as they don't bring so much bloat with them.

If feel

```cpp
for (bs::UINT32 i = 0; i < (bs::UINT32)vector.size(); i++) { ... }
```

is more readable than

```cpp
for (bs::UINT32 i = 0; i < static_cast<UINT32>(vector.size()); i++) { ... }
```

for example. I do agree that one should use `reinterpret_cast` to do pointer casting between objects though. But it might just be that I'm not as used to C++ style casts.

Reply:
I'm fine with that. I always use C++ style casts for everything as I find them to carry more information about the intent. But in the end it usually comes down to personal taste.